File: packages/python/hushspec/validate.py
from __future__ import annotations

import re
from dataclasses import dataclass, field

from datetime import date

from hushspec.extensions import DetectionLevel, TransitionTrigger
from hushspec.schema import Classification, HushSpec, LifecycleState
from hushspec.version import is_supported

_CAPABILITY_NAMES = frozenset(
    {"file_access", "file_write", "egress", "shell", "tool_call", "patch", "custom"}
)

_BUDGET_NAMES = frozenset(
    {"file_writes", "egress_calls", "shell_commands", "tool_calls", "patches", "custom_calls"}
)

_DURATION_PATTERN = re.compile(r"^\d+[smhd]$")
_DETECTION_LEVEL_ORDER = {
    DetectionLevel.SAFE: 0,
    DetectionLevel.SUSPICIOUS: 1,
    DetectionLevel.HIGH: 2,
    DetectionLevel.CRITICAL: 3,
}


@dataclass
class ValidationError:
    code: str
    message: str

    def __str__(self) -> str:
        return self.message


@dataclass
class ValidationResult:
    errors: list[ValidationError] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)

    @property
    def is_valid(self) -> bool:
        return not self.errors


def validate(spec: HushSpec) -> ValidationResult:
    errors: list[ValidationError] = []
    warnings: list[str] = []

    if not is_supported(spec.hushspec):
        errors.append(
            ValidationError("unsupported_version", f"unsupported hushspec version: {spec.hushspec}")
        )

    if spec.rules is not None:
        _validate_rules(spec.rules, errors)

        if (
            spec.rules.forbidden_paths is None
            and spec.rules.path_allowlist is None
            and spec.rules.egress is None
            and spec.rules.secret_patterns is None
            and spec.rules.patch_integrity is None
            and spec.rules.shell_commands is None
            and spec.rules.tool_access is None
            and spec.rules.computer_use is None
            and spec.rules.remote_desktop_channels is None
            and spec.rules.input_injection is None
        ):
            warnings.append("no rules configured")
    else:
        warnings.append("no rules section present")

    if spec.extensions is not None:
        _validate_posture(spec.extensions, errors, warnings)
        _validate_origins(spec.extensions, errors)
        _validate_detection(spec.extensions, errors, warnings)

    _validate_governance(spec, warnings)

    return ValidationResult(errors=errors, warnings=warnings)


def _validate_rules(rules: object, errors: list[ValidationError]) -> None:
    from hushspec.rules import Rules

    assert isinstance(rules, Rules)

    if rules.secret_patterns is not None:
        seen: set[str] = set()
        for pattern in rules.secret_patterns.patterns:
            if pattern.name in seen:
                errors.append(
                    ValidationError(
                        "duplicate_pattern_name",
                        f"duplicate secret pattern name: {pattern.name}",
                    )
                )
            seen.add(pattern.name)
            _validate_regex(
                pattern.pattern,
                f"secret_patterns.patterns.{pattern.name}",
                errors,
            )

    if rules.patch_integrity is not None:
        if rules.patch_integrity.max_imbalance_ratio <= 0.0:
            errors.append(
                ValidationError(
                    "invalid_ratio",
                    "rules.patch_integrity.max_imbalance_ratio must be > 0",
                )
            )
        for index, pattern in enumerate(rules.patch_integrity.forbidden_patterns):
            _validate_regex(
                pattern,
                f"rules.patch_integrity.forbidden_patterns[{index}]",
                errors,
            )

    if rules.shell_commands is not None:
        for index, pattern in enumerate(rules.shell_commands.forbidden_patterns):
            _validate_regex(
                pattern,
                f"rules.shell_commands.forbidden_patterns[{index}]",
                errors,
            )

    if rules.tool_access is not None and rules.tool_access.max_args_size == 0:
        errors.append(
            ValidationError(
                "invalid_max_args_size",
                "rules.tool_access.max_args_size must be >= 1",
            )
        )


def _validate_posture(ext: object, errors: list[ValidationError], warnings: list[str]) -> None:
    from hushspec.extensions import Extensions

    assert isinstance(ext, Extensions)

    if ext.posture is None:
        return

    posture = ext.posture

    if not posture.states:
        errors.append(
            ValidationError(
                "empty_states",
                "posture.states must define at least one state",
            )
        )

    if posture.initial not in posture.states:
        errors.append(
            ValidationError(
                "invalid_posture_initial",
                f"posture.initial '{posture.initial}' does not reference a defined state",
            )
        )

    for state_name, state in posture.states.items():
        for capability in state.capabilities:
            if capability not in _CAPABILITY_NAMES:
                warnings.append(
                    f"posture.states.{state_name}.capabilities includes unknown capability '{capability}'"
                )

        for budget_key, value in state.budgets.items():
            if value < 0:
                errors.append(
                    ValidationError(
                        "negative_budget",
                        f"posture.states.{state_name}.budgets.{budget_key} must be non-negative, got {value}",
                    )
                )
            if budget_key not in _BUDGET_NAMES:
                warnings.append(
                    f"posture.states.{state_name}.budgets uses unknown budget key '{budget_key}'"
                )

    for index, transition in enumerate(posture.transitions):
        if transition.from_state != "*" and transition.from_state not in posture.states:
            errors.append(
                ValidationError(
                    "invalid_transition_from",
                    f"posture.transitions[{index}].from '{transition.from_state}' does not reference a defined state",
                )
            )

        if transition.to == "*":
            errors.append(
                ValidationError(
                    "invalid_transition_to",
                    f"posture.transitions[{index}].to cannot be '*'",
                )
            )
        elif transition.to not in posture.states:
            errors.append(
                ValidationError(
                    "invalid_transition_to",
                    f"posture.transitions[{index}].to '{transition.to}' does not reference a defined state",
                )
            )

        if transition.on != TransitionTrigger.TIMEOUT:
            if transition.after is not None and not _is_valid_duration(transition.after):
                errors.append(
                    ValidationError(
                        "invalid_duration",
                        f"posture.transitions[{index}].after must match ^\\d+[smhd]$",
                    )
                )

        if transition.on == TransitionTrigger.TIMEOUT:
            if transition.after is None:
                errors.append(
                    ValidationError(
                        "missing_timeout_after",
                        f"posture.transitions[{index}]: timeout trigger requires 'after' field",
                    )
                )
            elif not _is_valid_duration(transition.after):
                errors.append(
                    ValidationError(
                        "invalid_duration",
                        f"posture.transitions[{index}].after must match ^\\d+[smhd]$",
                    )
                )


def _validate_origins(ext: object, errors: list[ValidationError]) -> None:
    from hushspec.extensions import Extensions

    assert isinstance(ext, Extensions)

    if ext.origins is None:
        return

    origins = ext.origins

    posture_states: set[str] | None = None
    if ext.posture is not None:
        posture_states = set(ext.posture.states)

    seen_ids: set[str] = set()
    for index, profile in enumerate(origins.profiles):
        if profile.id in seen_ids:
            errors.append(
                ValidationError(
                    "duplicate_origin_profile_id",
                    f"duplicate origin profile id: '{profile.id}'",
                )
            )
        seen_ids.add(profile.id)

        if profile.posture is not None:
            if posture_states is None:
                errors.append(
                    ValidationError(
                        "invalid_origin_posture",
                        f"origins.profiles[{index}].posture requires extensions.posture to be defined",
                    )
                )
            elif profile.posture not in posture_states:
                errors.append(
                    ValidationError(
                        "invalid_origin_posture",
                        f"origins.profiles[{index}].posture '{profile.posture}' does not reference a defined posture state",
                    )
                )


def _validate_detection(
    ext: object, errors: list[ValidationError], warnings: list[str]
) -> None:
    from hushspec.extensions import Extensions

    assert isinstance(ext, Extensions)

    if ext.detection is None:
        return

    detection = ext.detection

    if detection.prompt_injection is not None:
        pi = detection.prompt_injection

        if pi.max_scan_bytes == 0:
            errors.append(
                ValidationError(
                    "invalid_max_scan_bytes",
                    "detection.prompt_injection.max_scan_bytes must be >= 1",
                )
            )

        warn_level = pi.warn_at_or_above or DetectionLevel.SUSPICIOUS
        block_level = pi.block_at_or_above or DetectionLevel.HIGH
        if _DETECTION_LEVEL_ORDER[block_level] < _DETECTION_LEVEL_ORDER[warn_level]:
            warnings.append(
                "detection.prompt_injection: block_at_or_above is less strict than warn_at_or_above"
            )

    if detection.jailbreak is not None:
        jb = detection.jailbreak

        if jb.block_threshold is not None and jb.block_threshold > 100:
            errors.append(
                ValidationError(
                    "out_of_range",
                    "detection.jailbreak.block_threshold must be between 0 and 100",
                )
            )
        if jb.warn_threshold is not None and jb.warn_threshold > 100:
            errors.append(
                ValidationError(
                    "out_of_range",
                    "detection.jailbreak.warn_threshold must be between 0 and 100",
                )
            )
        if jb.max_input_bytes == 0:
            errors.append(
                ValidationError(
                    "invalid_max_input_bytes",
                    "detection.jailbreak.max_input_bytes must be >= 1",
                )
            )

        block_threshold = jb.block_threshold if jb.block_threshold is not None else 80
        warn_threshold = jb.warn_threshold if jb.warn_threshold is not None else 50
        if block_threshold < warn_threshold:
            warnings.append(
                "detection.jailbreak: block_threshold is lower than warn_threshold"
            )

    if detection.threat_intel is not None:
        ti = detection.threat_intel

        if ti.similarity_threshold is not None:
            if not (0.0 <= ti.similarity_threshold <= 1.0):
                errors.append(
                    ValidationError(
                        "out_of_range",
                        "detection.threat_intel.similarity_threshold must be between 0.0 and 1.0",
                    )
                )

        if ti.top_k == 0:
            errors.append(
                ValidationError(
                    "invalid_top_k",
                    "detection.threat_intel.top_k must be >= 1",
                )
            )


# Pattern that detects regex features outside the RE2 subset.
#
# HushSpec requires all regex patterns to be RE2-compatible to prevent ReDoS
# attacks. Python's ``re`` module uses a backtracking engine that is vulnerable
# to catastrophic backtracking with certain pattern constructs.  By restricting
# patterns to the RE2 subset we ensure safe O(mn) evaluation across all SDKs.
#
# Disallowed features:
# - Backreferences: \1, \2, ..., \k<name>
# - Lookahead: (?=...), (?!...)
# - Lookbehind: (?<=...), (?<!...)
# - Atomic groups: (?>...)
# - Possessive quantifiers: *+, ++, ?+
# - Conditional patterns: (?(...)...|...)
# - Recursive patterns: (?R), (?1), (?2), ...
# - Named backreferences: (?P=name)
# - Subroutine calls: \g<name>
_RE2_DISALLOWED = re.compile(
    r"\\[1-9]|\\k<|\(\?[=!]|\(\?<[=!]|\(\?>|\*\+|\+\+|\?\+|\(\?\(|\(\?R\)|\(\?\d+\)|\(\?P=|\\g<"
)


def is_safe_regex(pattern: str) -> bool:
    """Check whether a regex pattern is safe for evaluation (RE2-compatible).

    Returns ``True`` if the pattern uses only RE2-compatible features.
    Returns ``False`` if the pattern contains backreferences, lookaround,
    atomic groups, possessive quantifiers, or other non-RE2 features.
    """
    return _RE2_DISALLOWED.search(pattern) is None


def _validate_regex(pattern: str, path: str, errors: list[ValidationError]) -> None:
    try:
        re.compile(pattern)
    except re.error as e:
        errors.append(
            ValidationError(
                "invalid_regex",
                f"{path} must be a valid regular expression: {e}",
            )
        )
        return

    if not is_safe_regex(pattern):
        errors.append(
            ValidationError(
                "non_re2_regex",
                f"{path}: pattern uses features not in the RE2 subset "
                "(backreferences, lookaround, etc.) which may cause ReDoS",
            )
        )


def _is_valid_duration(value: str) -> bool:
    return bool(_DURATION_PATTERN.match(value))


def _validate_governance(spec: HushSpec, warnings: list[str]) -> None:
    if spec.metadata is None:
        return

    metadata = spec.metadata

    if metadata.lifecycle_state is not None:
        if metadata.lifecycle_state in (LifecycleState.DEPRECATED, LifecycleState.ARCHIVED):
            warnings.append(
                f"policy lifecycle state is '{metadata.lifecycle_state.value}'"
            )

    if metadata.expiry_date is not None:
        today = date.today().isoformat()
        if metadata.expiry_date < today:
            warnings.append(
                f"policy expiry_date '{metadata.expiry_date}' is in the past"
            )

    if metadata.approved_by is not None and metadata.approval_date is None:
        warnings.append("approved_by is set but approval_date is missing")

    if metadata.classification == Classification.RESTRICTED and metadata.approved_by is None:
        warnings.append("classification is 'restricted' but no approved_by is set")

File: packages/go/hushspec/validate.go
package hushspec

import (
	"fmt"
	"regexp"
	"time"
)

type ValidationResult struct {
	Errors   []ValidationError
	Warnings []string
}

type ValidationError struct {
	Code    string
	Message string
}

func (r *ValidationResult) IsValid() bool {
	return len(r.Errors) == 0
}

func (r *ValidationResult) addError(code, msg string) {
	r.Errors = append(r.Errors, ValidationError{Code: code, Message: msg})
}

func (r *ValidationResult) addWarning(msg string) {
	r.Warnings = append(r.Warnings, msg)
}

// Validate checks version support, cross-field constraints, regex syntax,
// and extension consistency. Returns errors and advisory warnings.
func Validate(spec *HushSpec) *ValidationResult {
	result := &ValidationResult{}

	if spec.HushSpecVersion == "" {
		result.addError("MISSING_VERSION", "missing or empty 'hushspec' version field")
	} else if !IsSupported(spec.HushSpecVersion) {
		result.addError("UNSUPPORTED_VERSION",
			fmt.Sprintf("unsupported HushSpec version %q; supported: %v", spec.HushSpecVersion, SupportedVersions))
	}

	if spec.MergeStrategy != "" && !containsTyped(spec.MergeStrategy, MergeStrategies) {
		result.addError("INVALID_MERGE_STRATEGY",
			fmt.Sprintf("invalid merge_strategy %q; must be one of: replace, merge, deep_merge", spec.MergeStrategy))
	}

	if spec.Rules != nil {
		validateRules(spec.Rules, result)
	}
	if spec.Extensions != nil {
		validateExtensions(spec.Extensions, result)
	}

	validateGovernance(spec, result)

	return result
}

func validateGovernance(spec *HushSpec, result *ValidationResult) {
	if spec.Metadata == nil {
		return
	}
	m := spec.Metadata

	if m.LifecycleState == LifecycleStateDeprecated || m.LifecycleState == LifecycleStateArchived {
		result.addWarning(fmt.Sprintf("policy lifecycle state is '%s'", m.LifecycleState))
	}

	if m.ExpiryDate != "" {
		today := currentDateISO()
		if m.ExpiryDate < today {
			result.addWarning(fmt.Sprintf("policy expiry_date '%s' is in the past", m.ExpiryDate))
		}
	}

	if m.ApprovedBy != "" && m.ApprovalDate == "" {
		result.addWarning("approved_by is set but approval_date is missing")
	}

	if m.Classification == ClassificationRestricted && m.ApprovedBy == "" {
		result.addWarning("classification is 'restricted' but no approved_by is set")
	}
}

func validateRules(rules *Rules, result *ValidationResult) {
	if rules.SecretPatterns != nil {
		seen := make(map[string]bool)
		for _, pattern := range rules.SecretPatterns.Patterns {
			if pattern.Name == "" {
				result.addError("EMPTY_PATTERN_NAME", "secret pattern has an empty name")
				continue
			}
			if seen[pattern.Name] {
				result.addError("DUPLICATE_PATTERN_NAME",
					fmt.Sprintf("duplicate secret pattern name %q", pattern.Name))
			}
			seen[pattern.Name] = true
			if !containsTyped(string(pattern.Severity), Severities) {
				result.addError("INVALID_SEVERITY",
					fmt.Sprintf("secret_patterns.patterns.%s.severity %q must be critical, error, or warn", pattern.Name, pattern.Severity))
			}
			validateRegex(pattern.Pattern, fmt.Sprintf("secret_patterns.patterns.%s", pattern.Name), result)
		}
	}

	if rules.Egress != nil {
		switch {
		case rules.Egress.Default == "":
			result.addWarning("egress rule has no default action specified; will default to block")
		case !containsTyped(rules.Egress.Default, DefaultActions):
			result.addError("INVALID_DEFAULT_ACTION",
				fmt.Sprintf("egress default action %q must be 'allow' or 'block'", rules.Egress.Default))
		}
	}

	if rules.ToolAccess != nil {
		switch {
		case rules.ToolAccess.Default == "":
			result.addWarning("tool_access rule has no default action specified; will default to allow")
		case !containsTyped(rules.ToolAccess.Default, DefaultActions):
			result.addError("INVALID_DEFAULT_ACTION",
				fmt.Sprintf("tool_access default action %q must be 'allow' or 'block'", rules.ToolAccess.Default))
		}
		if rules.ToolAccess.MaxArgsSize != nil && *rules.ToolAccess.MaxArgsSize < 1 {
			result.addError("INVALID_MAX_ARGS_SIZE", "rules.tool_access.max_args_size must be >= 1")
		}
	}

	if rules.ComputerUse != nil && rules.ComputerUse.Mode != "" && !containsTyped(rules.ComputerUse.Mode, ComputerUseModes) {
		result.addError("INVALID_COMPUTER_USE_MODE",
			fmt.Sprintf("rules.computer_use.mode %q must be observe, guardrail, or fail_closed", rules.ComputerUse.Mode))
	}

	if rules.PatchIntegrity != nil {
		if rules.PatchIntegrity.MaxAdditions < 0 {
			result.addError("NEGATIVE_LIMIT", "patch_integrity max_additions must be non-negative")
		}
		if rules.PatchIntegrity.MaxDeletions < 0 {
			result.addError("NEGATIVE_LIMIT", "patch_integrity max_deletions must be non-negative")
		}
		if rules.PatchIntegrity.MaxImbalanceRatio != nil && *rules.PatchIntegrity.MaxImbalanceRatio <= 0 {
			result.addError("INVALID_RATIO", "patch_integrity max_imbalance_ratio must be > 0")
		}
		for index, pattern := range rules.PatchIntegrity.ForbiddenPatterns {
			validateRegex(pattern, fmt.Sprintf("rules.patch_integrity.forbidden_patterns[%d]", index), result)
		}
	}

	if rules.ShellCommands != nil {
		for index, pattern := range rules.ShellCommands.ForbiddenPatterns {
			validateRegex(pattern, fmt.Sprintf("rules.shell_commands.forbidden_patterns[%d]", index), result)
		}
	}
}

func validateExtensions(ext *Extensions, result *ValidationResult) {
	if ext.Posture != nil {
		validatePosture(ext.Posture, result)
	}
	if ext.Origins != nil {
		validateOrigins(ext, result)
	}
	if ext.Detection != nil {
		validateDetection(ext.Detection, result)
	}
}

func validatePosture(posture *PostureExtension, result *ValidationResult) {
	if len(posture.States) == 0 {
		result.addError("EMPTY_STATES", "posture.states must define at least one state")
	}

	if posture.Initial == "" {
		result.addError("MISSING_INITIAL_STATE", "posture.initial is required")
	} else if _, ok := posture.States[posture.Initial]; !ok {
		result.addError("INVALID_INITIAL_STATE",
			fmt.Sprintf("posture.initial %q does not reference a defined state", posture.Initial))
	}

	for stateName, state := range posture.States {
		for _, capability := range state.Capabilities {
			if !isKnownCapability(capability) {
				result.addWarning(
					fmt.Sprintf("posture.states.%s.capabilities includes unknown capability %q", stateName, capability),
				)
			}
		}
		for budgetKey, value := range state.Budgets {
			if value < 0 {
				result.addError("NEGATIVE_BUDGET",
					fmt.Sprintf("posture.states.%s.budgets.%s must be non-negative, got %d", stateName, budgetKey, value))
			}
			if !isKnownBudgetKey(budgetKey) {
				result.addWarning(
					fmt.Sprintf("posture.states.%s.budgets uses unknown budget key %q", stateName, budgetKey),
				)
			}
		}
	}

	for index, transition := range posture.Transitions {
		if transition.From != "*" {
			if _, ok := posture.States[transition.From]; !ok {
				result.addError("INVALID_TRANSITION_STATE",
					fmt.Sprintf("posture.transitions[%d].from %q does not reference a defined state", index, transition.From))
			}
		}

		if transition.To == "*" {
			result.addError("INVALID_TRANSITION_STATE",
				fmt.Sprintf("posture.transitions[%d].to cannot be '*'", index))
		} else if _, ok := posture.States[transition.To]; !ok {
			result.addError("INVALID_TRANSITION_STATE",
				fmt.Sprintf("posture.transitions[%d].to %q does not reference a defined state", index, transition.To))
		}

		if !containsTyped(transition.On, TransitionTriggers) {
			result.addError("INVALID_TRANSITION_TRIGGER",
				fmt.Sprintf("posture.transitions[%d].on %q is not a valid trigger", index, transition.On))
		}

		if transition.On == TransitionTriggerTimeout {
			if transition.After == nil {
				result.addError("MISSING_TIMEOUT_AFTER",
					fmt.Sprintf("posture.transitions[%d]: timeout trigger requires 'after' field", index))
			} else if !isValidDuration(*transition.After) {
				result.addError("INVALID_DURATION",
					fmt.Sprintf("posture.transitions[%d].after must match ^\\d+[smhd]$", index))
			}
		} else if transition.After != nil && !isValidDuration(*transition.After) {
			result.addError("INVALID_DURATION",
				fmt.Sprintf("posture.transitions[%d].after must match ^\\d+[smhd]$", index))
		}
	}
}

func validateOrigins(ext *Extensions, result *ValidationResult) {
	origins := ext.Origins
	seen := make(map[string]bool)
	postureStates := map[string]bool{}
	if ext.Posture != nil {
		for stateName := range ext.Posture.States {
			postureStates[stateName] = true
		}
	}

	if origins.DefaultBehavior != nil && !containsTyped(*origins.DefaultBehavior, OriginDefaultBehaviors) {
		result.addError("INVALID_DEFAULT_BEHAVIOR",
			fmt.Sprintf("origins.default_behavior %q must be 'deny' or 'minimal_profile'", *origins.DefaultBehavior))
	}

	for index, profile := range origins.Profiles {
		if profile.ID == "" {
			result.addError("EMPTY_ORIGIN_ID", "origin profile has an empty id")
			continue
		}
		if seen[profile.ID] {
			result.addError("DUPLICATE_ORIGIN_ID",
				fmt.Sprintf("duplicate origin profile id %q", profile.ID))
		}
		seen[profile.ID] = true

		if profile.Match != nil {
			if profile.Match.SpaceType != "" && !containsTyped(profile.Match.SpaceType, OriginSpaceTypes) {
				result.addError("INVALID_ORIGIN_SPACE_TYPE",
					fmt.Sprintf("origins.profiles[%d].match.space_type %q is not valid", index, profile.Match.SpaceType))
			}
			if profile.Match.Visibility != "" && !containsTyped(profile.Match.Visibility, OriginVisibilities) {
				result.addError("INVALID_ORIGIN_VISIBILITY",
					fmt.Sprintf("origins.profiles[%d].match.visibility %q is not valid", index, profile.Match.Visibility))
			}
		}

		if profile.Posture != nil {
			if len(postureStates) == 0 {
				result.addError("INVALID_ORIGIN_POSTURE",
					fmt.Sprintf("origins.profiles[%d].posture requires extensions.posture to be defined", index))
			} else if !postureStates[*profile.Posture] {
				result.addError("INVALID_ORIGIN_POSTURE",
					fmt.Sprintf("origins.profiles[%d].posture %q does not reference a defined posture state", index, *profile.Posture))
			}
		}

		if profile.Budgets != nil {
			validateOptionalNonNegativeInt(profile.Budgets.ToolCalls, "NEGATIVE_BUDGET",
				fmt.Sprintf("origins.profiles[%d].budgets.tool_calls must be non-negative", index), result)
			validateOptionalNonNegativeInt(profile.Budgets.EgressCalls, "NEGATIVE_BUDGET",
				fmt.Sprintf("origins.profiles[%d].budgets.egress_calls must be non-negative", index), result)
			validateOptionalNonNegativeInt(profile.Budgets.ShellCommands, "NEGATIVE_BUDGET",
				fmt.Sprintf("origins.profiles[%d].budgets.shell_commands must be non-negative", index), result)
		}

		if profile.Bridge != nil {
			for targetIndex, target := range profile.Bridge.AllowedTargets {
				if target.SpaceType != "" && !containsTyped(target.SpaceType, OriginSpaceTypes) {
					result.addError("INVALID_BRIDGE_SPACE_TYPE",
						fmt.Sprintf("origins.profiles[%d].bridge.allowed_targets[%d].space_type %q is not valid", index, targetIndex, target.SpaceType))
				}
				if target.Visibility != "" && !containsTyped(target.Visibility, OriginVisibilities) {
					result.addError("INVALID_BRIDGE_VISIBILITY",
						fmt.Sprintf("origins.profiles[%d].bridge.allowed_targets[%d].visibility %q is not valid", index, targetIndex, target.Visibility))
				}
			}
		}
	}
}

func validateDetection(detection *DetectionExtension, result *ValidationResult) {
	if detection.PromptInjection != nil {
		prompt := detection.PromptInjection
		if prompt.WarnAtOrAbove != nil && !containsTyped(*prompt.WarnAtOrAbove, DetectionLevels) {
			result.addError("INVALID_DETECTION_LEVEL",
				fmt.Sprintf("detection.prompt_injection.warn_at_or_above %q is not valid", *prompt.WarnAtOrAbove))
		}
		if prompt.BlockAtOrAbove != nil && !containsTyped(*prompt.BlockAtOrAbove, DetectionLevels) {
			result.addError("INVALID_DETECTION_LEVEL",
				fmt.Sprintf("detection.prompt_injection.block_at_or_above %q is not valid", *prompt.BlockAtOrAbove))
		}
		if prompt.MaxScanBytes != nil && *prompt.MaxScanBytes < 1 {
			result.addError("INVALID_MAX_SCAN_BYTES", "detection.prompt_injection.max_scan_bytes must be >= 1")
		}

		warnLevel := DetectionLevelSuspicious
		if prompt.WarnAtOrAbove != nil {
			warnLevel = *prompt.WarnAtOrAbove
		}
		blockLevel := DetectionLevelHigh
		if prompt.BlockAtOrAbove != nil {
			blockLevel = *prompt.BlockAtOrAbove
		}
		if containsTyped(warnLevel, DetectionLevels) && containsTyped(blockLevel, DetectionLevels) && detectionRank(blockLevel) < detectionRank(warnLevel) {
			result.addWarning("detection.prompt_injection: block_at_or_above is less strict than warn_at_or_above")
		}
	}

	if detection.Jailbreak != nil {
		jailbreak := detection.Jailbreak
		if jailbreak.BlockThreshold != nil && (*jailbreak.BlockThreshold < 0 || *jailbreak.BlockThreshold > 100) {
			result.addError("INVALID_BLOCK_THRESHOLD", "detection.jailbreak.block_threshold must be between 0 and 100")
		}
		if jailbreak.WarnThreshold != nil && (*jailbreak.WarnThreshold < 0 || *jailbreak.WarnThreshold > 100) {
			result.addError("INVALID_WARN_THRESHOLD", "detection.jailbreak.warn_threshold must be between 0 and 100")
		}
		if jailbreak.MaxInputBytes != nil && *jailbreak.MaxInputBytes < 1 {
			result.addError("INVALID_MAX_INPUT_BYTES", "detection.jailbreak.max_input_bytes must be >= 1")
		}

		blockThreshold := 80
		if jailbreak.BlockThreshold != nil {
			blockThreshold = *jailbreak.BlockThreshold
		}
		warnThreshold := 50
		if jailbreak.WarnThreshold != nil {
			warnThreshold = *jailbreak.WarnThreshold
		}
		if blockThreshold < warnThreshold {
			result.addWarning("detection.jailbreak: block_threshold is lower than warn_threshold")
		}
	}

	if detection.ThreatIntel != nil {
		threatIntel := detection.ThreatIntel
		if threatIntel.SimilarityThreshold != nil {
			if *threatIntel.SimilarityThreshold < 0.0 || *threatIntel.SimilarityThreshold > 1.0 {
				result.addError("THRESHOLD_OUT_OF_RANGE",
					"detection.threat_intel.similarity_threshold must be between 0.0 and 1.0")
			}
		}
		if threatIntel.TopK != nil && *threatIntel.TopK < 1 {
			result.addError("INVALID_TOP_K", "detection.threat_intel.top_k must be >= 1")
		}
	}
}

func validateOptionalNonNegativeInt(value *int, code, msg string, result *ValidationResult) {
	if value != nil && *value < 0 {
		result.addError(code, msg)
	}
}

// validateRegex rejects non-RE2 patterns. Go's regexp is RE2-only, so any
// pattern that compiles is inherently ReDoS-safe.
func validateRegex(pattern, path string, result *ValidationResult) {
	if _, err := regexp.Compile(pattern); err != nil {
		result.addError("INVALID_REGEX",
			fmt.Sprintf("%s must be a valid regular expression: %v", path, err))
	}
}

func isKnownCapability(value string) bool {
	switch value {
	case "file_access", "file_write", "egress", "shell", "tool_call", "patch", "custom":
		return true
	default:
		return false
	}
}

func isKnownBudgetKey(value string) bool {
	switch value {
	case "file_writes", "egress_calls", "shell_commands", "tool_calls", "patches", "custom_calls":
		return true
	default:
		return false
	}
}

func isValidDuration(value string) bool {
	matched, _ := regexp.MatchString(`^\d+[smhd]$`, value)
	return matched
}

func containsTyped[T comparable](value T, allowed map[T]struct{}) bool {
	_, ok := allowed[value]
	return ok
}

func detectionRank(value DetectionLevel) int {
	switch value {
	case DetectionLevelSafe:
		return 0
	case DetectionLevelSuspicious:
		return 1
	case DetectionLevelHigh:
		return 2
	case DetectionLevelCritical:
		return 3
	default:
		return -1
	}
}

func currentDateISO() string {
	return time.Now().UTC().Format("2006-01-02")
}
